### ISSUE TITLE
Sample_weights - question about what happens, improve documentation

### ISSUE THREAD

**Issue by joseortiz3** [EpistasisLab/tpot#708]
After sample_weights were introduced in response to [this](https://github.com/EpistasisLab/tpot/issues/310), the documentation for sample_weights was updated with this description:

> Per-sample weights. Higher weights force TPOT to put more emphasis on those points.

This leaves some questions unanswered, and I want to make sure I understand what is happening "under the hood". 

From looking at [operator_utils.py]..........(https://github.com/EpistasisLab/tpot/blob/master/tpot/operator_utils.py),

```python
def set_sample_weight(pipeline_steps, sample_weight=None):
    """Recursively iterates through all objects in the pipeline and sets sample weight.
    Parameters
    ----------
    pipeline_steps: array-like
        List of (str, obj) tuples from a scikit-learn pipeline or related object
    sample_weight: array-like
        List of sample weight
    Returns
    -------
    sample_weight_dict:
        A dictionary of sample_weight
    """
    sample_weight_dict = {}
    if not isinstance(sample_weight, type(None)):
        for (pname, obj) in pipeline_steps:
            if inspect.getargspec(obj.fit).args.count('sample_weight'):
                step_sw = pname + '__sample_weight'
                sample_weight_dict[step_sw] = sample_weight

    if sample_weight_dict:
        return sample_weight_dict
    else:
        return None
```
 ......it seems that each element in the pipeline is checked to see if it has a `.fit()` method that accepts a sample_weight argument.

If so, I have two questions regarding "missing opportunities" to pass sample_weights.

1) it might miss passing sample_weights to `fit_transform()` (but hopefully tpot never uses `fit_transform` and instead simply calls `fit` and `transform` separately)
2) it might miss passing sample_weights to some function that could use it, although I don't know of any (and there might not be any)

Also I wanted to know if sample_weights is used in tpot's scoring function, which determines preferences between pipelines?

Finally, maybe another sentence or two in the documentation to explain briefly what is happening with sample_weights would be helpful to future people having these same questions.

Thanks for tpot! Wonderful seeing this project progress over the last two years.

**Comment by weixuanfu** [comment]
Thank you for your idea about sample weight.

For 2 questions, I think all the operators in TPOT should follow the scikit-learn API which means that they should has `fit`. But I will double-check it.

**Comment by rhiever** [comment]
AFAIK no operator in sklearn uses `sample_weights` without having a `fit` function. In fact, we couldn't use an operator in sklearn if it didn't have a `fit` function because it wouldn't work in a sklearn pipeline. Furthermore, because we use sklearn pipelines as our solution representation, sklearn pipeline only use `fit` and `prediction`/`transform` separately.

>Also I wanted to know if sample_weights is used in tpot's scoring function, which determines preferences between pipelines?

No, the `sample_weights` are not used in the scoring function unless you separately tell TPOT to use a scoring function that takes sample weights into account. Even in sklearn, setting `sample_weights` affects only the training process of the algorithms, not their evaluation.

👍 to expanding the docs to expand the description of `sample_weights` to address some of the questions in this thread. Are you willing to submit that PR, @joseortiz3?

**Comment by joseortiz3** [comment]
I can submit a PR for the documentation, now that you guys have explained it to me.

Now I need to figure out how to change the TPOT scoring function to use sample_weights. Before I begin this journey, any tips?

**Comment by rhiever** [comment]
Thank you, @joseortiz3. We look forward to reviewing the docs PR.

TPOT provides access to a `scoring` parameter when creating a TPOT instance. With that parameter, you can use any [built-in sklearn metric](http://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) or even custom metrics (see the [TPOT API](https://epistasislab.github.io/tpot/api/) for details). For example, the [accuracy_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score) function in sklearn takes a `sample_weight` parameter.

**Comment by joseortiz3** [comment]
Also I found that there currently is no check for invalid sample_weights. If there is a single `NaN`, tpot fails later in `_update_top_pipeline`, when it finds that all the pipelines have `-inf` scores.

**Comment by josiahjohnston** [comment]
@joseortiz3 Did you ever figure out how to pass per-sample weights into a scoring function?

In my regression problem, most of the per-sample weights are determined by information that is not part of y, so I can't derive sample weights from available info inside the scoring function. The only robust way I can think of would require hacking tpot.gp_deap & sklearn's _fit_and_score which seems less than ideal. 
...Hmmm. Sounds like I'm re-suggesting a strategy @weixuanfu proposed in https://github.com/EpistasisLab/tpot/issues/310
